README.md
Stats
=====
A module for collecting runtime stats and presenting (formatting) them properly.

```sh
npm install https://github.com/Perennials/stats-node/tarball/master
```

<!-- MarkdownTOC -->

- [API](#api)
	- [Stats](#stats)
		- [Methods](#methods)
			- [.startTimer()](#starttimer)
			- [.endTimer()](#endtimer)
			- [.saveTimer()](#savetimer)
			- [.addTimer()](#addtimer)
			- [.setStat()](#setstat)
			- [.addStat()](#addstat)
			- [.getStat()](#getstat)
			- [.setUnit()](#setunit)
			- [.getUnit()](#getunit)
			- [.getCount()](#getcount)
			- [.getAverage()](#getaverage)
			- [.saveAverage()](#saveaverage)
			- [.getStats()](#getstats)
	- [FormatUtils](#formatutils)
		- [Methods](#methods-1)
			- [FormatUtils.formatBytes()](#formatutilsformatbytes)
	- [TimeUtils](#timeutils)
		- [Methods](#methods-2)
			- [.timeAgoFormat()](#timeagoformat)
			- [.timeAgo()](#timeago)
- [Authors](#authors)

<!-- /MarkdownTOC -->


API
---

### Stats
A class that collects runtime stats, each having a unique name.

Stats can have an unit assigned to them so they can be formatted accordingly.
Formatting functionality is provided for time and byte units.

```js
var Stats = require( 'Stats' );
```

Nesting and summing timers:

```js
var stats = new Stats();
stats.startTimer();
for ( ... ) {
	func1();
	stats.startTimer();
	func2();
	// measure the sum of the time of all calls to func2()
	stats.addTimer( 'My.Inner.Thingie' );
}
// measure the time taken for the for last loop (not the sum)
stats.saveTimer( 'My.Outer.Loop' );

console.log( stats.getStats() );
```

Timers and counting:

```js
var stats = new Stats();
var slowestSQL = null;
var slowestTime = 0;
for ( ... ) {
	stats.startTimer();
	perform_sql_query( sql );
	var last = stats.endTimer();
	if ( last > slowestTime ) {
		slowestTime = last;
		slowestSQL = sql;
	}
	stats.addStat( 'SQL.Queries.Count', 1 );
}
stats.setStat( 'SQL.Slowest.Query', slowestSQL );
stats.setStat( 'SQL.Slowest.Timing', slowestTime );

console.log( stats.getStats() );
```

Named, async timers:

```js
var stats = new Stats();
var sql = new Sql();

// we can use regular names (string, int) or objects
stats.startTimer( sql );
sql.execute( function () {
	stats.saveTimer( sql, 'SQL.Timing' );
	stats.setStat( 'SQL.Text', sql.getText() );

	console.log( stats.getStats() );
} );
```

#### Methods

##### .startTimer()
Starts a new timer.

Nameless timers are stacked. I.e. you can call `startTimer()` again before the
previous timer was flushed with `endTimer()` Each `startTimer()` call must be
matched by a call to [`endTimer()`](#endtimer) or [`saveTimer()`](#savetimer).
This can be used only for synchronous applications.

Timers have millisecond precision.

Returns the timestamp when the timer was started (`Date.now()`).

```js
.startTimer(
	name:String|Number|Object|undefined
) : Number;
```

##### .endTimer()
Stops the timer started by the last [`startTimer()`](#starttimer) call.

Returns the elapsed time in milliseconds.

```js
.stopTimer(
	name:String|Number|Object|undefined
) : Number;
```

##### .saveTimer()
Stops the timer started by the last [`startTimer()`](#starttimer) call and saves its value to the list of stats.

`stat` is the name of the stat under which the timer value will be saved.
The `name` parameter is passed directly to `endTimer()` and the return value is the same.

```js
.saveTimer(
	name:String|Number|Object,
	stat:String
) : Number;
```

```js
.saveTimer(
	stat:String
) : Number;
```

##### .addTimer()
Stops the timer started by the last [`startTimer()`](#starttimer) call and adds its value to existing value in the list of stats.

`stat` is the name of the stat under which the timer value will be saved (added).
The `name` parameter is passed directly to `endTimer()` and the return value is the same.

```js
.addTimer(
	name:String|Number|Object,
	stat:String
) : Number;
```

```js
.addTimer(
	stat:String
) : Number;
```

##### .setStat()
Writes a value in the list of maintained stats.

If such stat already exists it is replaced.
`stat` is the name of the stat under which the timer value will be saved.
`unit` is optional unit of the value. The list of known units that will be
formatted properly includes `b` for bytes and `ms` for millliseconds.

```js
.setStat(
	stat:String,
	value:mixed,
	unit:String|undefined
) : mixed;
```

##### .addStat()
Adds to the value of existing property in the list of maintained stats.

If the property doesn't exist this is the identical to
[`setStat()`](#setstat). This should be called only on scalable (i.e. int,
float) stats. The also maintains the 'count' of how many times it was called,
i.e. how many times a value was added, making possible to find the average
(see [`getCount()`](#getcount)).

`stat` is the name of the stat under which the timer value will be saved.
`unit` is optional unit of the value. The list of known units that will be
formatted properly includes `b` for bytes and `ms` for millliseconds.

```js
.addStat(
	stat:String,
	value:mixed,
	unit:String|undefined
) : mixed;
```

##### .getStat()
Retrieves the value of stored property in the list of maintained stats.

```js
.addStat(
	stat:String
) : mixed|undefined;
```

##### .setUnit()
Sets (or changes) the unit of a property in the list of maintained stats.

```js
.addStat(
	stat:String,
	unit:String
) : String;
```

##### .getUnit()
Retrieves the unit of a property.

```js
.getUnit(
	stat:String
) : String|undefined;
```

##### .getCount()
Retrieves the 'count' of a stat in the list of maintained stats.

'count' means how many times [`addStat()`](#addstat) was called for this
property. The 'count' for stats created by [`setStat()`](#setstat) is 1.

Returns 0 if the property doesn't exist.

```js
.getCount(
	stat:String
) : Number;
```

##### .getAverage()
Retrieves the average for a stat.

This is for stats that are made of several [`addStat()`](#addstat) calls.

```js
.getAverage(
	stat:String
) : Number|undefined;
```

##### .saveAverage()
Saves the average value of a stat in a separate stat.

This is for stats that are made of several [`addStat()`](#addstat) calls.

```js
.saveAverage(
	stat:String,
	newname:String
) : Number|undefined;
```

##### .getStats()
Retrieves the list of stats.

Stats that have an unit specified are formated accordingly to be more readable.

Supported units are:

- `b`: bytes, formated as "1.5 MB", "2.12 KB", etc.
- `ms`: milliseconds, formated as "1.123456 s", "3 minutes 1 second", "5 days 3 hours".
- other units are just appended with a space to the value.

Returns a mapping where the key is the name of the stat and the value is its value.

```js
.getStats(
	options:Object|undefined
) : Object;
```

`options` can be given to specify the formating of known units. It can be an object like

```js
{
	Bytes: Object,
	Time: Object
}
```

The defaults are comming from `Stats.DefaultFormatting`. `Bytes` is passed to
`FormatUtils.formatBytes()` and `Time` is passed to
`TimeUtils.timeAgoFormat()`.

### FormatUtils
A collection of static methods for formatting units and similar.
This is used internally by the `Stats` class.

#### Methods

##### FormatUtils.formatBytes()
Converts a size in bytes to a bigger scale of bytes (e.g. KB, MB, etc). The
scale will be automatically determined.

```js
FormatUtils.formatBytes(
	bytes:Number,
	options:Object|undefined
) : String;
```

`options` takes the following format:

```js
{
	// sprintf style format string, defaults to FormatUtils.DefaultFormat.Bytes
	// it will receive two arguments- one float and one string for the unit.
	// The default format is '%01.2f %s'
	Format: String,

	// force an output unit
	// for IEC it could be 'B', 'KiB', 'MiB', 'GiB', 'TiB', 'PiB'
	// for SI it could be 'B', 'KB', 'MB', 'GB', 'TB', 'PB'
	ForceUnit: Boolean,

	// whether to use SI (1000 bytes in KB) units or IEC (1024 bytes in a Kib)
	// this will be infered from ForceUnit
	UseSi: Boolean

}
```

### TimeUtils
A collection of static methods for formatting time.
This is used internally by the `Stats` class.

#### Methods

##### .timeAgoFormat()
Returns the time difference between two dates as formatted string.

For example `1 year 2 months`, `1 minute 3 seconds`, `10 hours`.

If `date1` is omitted the number of milliseconds repsented by `date2` will be
treated as the difference between `date2` and `date1`.

```js
TimeUtils.timeAgoFormat(
	date2:Number|Date,
	date1:Number|Date|undefined,
	options:Object
) : String;
```

`options` takes the following format:

```js
{
	// If false the return value will contain only the most significant time unit (i.e. short format).
	Precise: Boolean,

	// sprintf format string for long format. The default is '%d %s %d %s'.
	LongFormat: String,

	// sprintf format string for short format. The default is '%d %s'.
	ShortFormat: String,

	// This provides support for translating the time interval names.
	// The defaults are 'year', 'month', 'day', etc. If 'short' is passed the default will be 'y', 'm', 'd', etc.
	Strings: Object|'short'
}
```

##### .timeAgo()
Returns the time difference between two dates as array.

For example `[1, 'year', 2, 'months']`, `[1, 'minute', 2, 'seconds']`, `[10, 'hours']`.

If `date1` is omitted the number of milliseconds repsented by `date2` will be
treated as the difference between `date2` and `date1`.

```js
TimeUtils.timeAgo(
	date2:Number|Date,
	date1:Number|Date|undefined,
	options:Object
) : String;
```

`options` takes the following format:

```js
{
	// If false the return value will contain only the most significant time unit (i.e. short format).
	Precise: Boolean
}
```


Authors
-------
Borislav Peev (borislav.asdf at gmail dot com)


CHANGELOG.md
1.1.1
---
- Fix crash when time name is number.

1.1
---
- `TimeUtils.timeAgo()` and `TimeUtils.timeAgoFormat()` now accept `Date`s.
- `TimeUtils.timeAgoFormat()` can now have customisable interval names.
- **Breaking:** `TimeUtils.timeAgoFormat()` now accepts its options as an `Object`.

1.0
---
- Initial port from PHP.

tests/tests.js
"use strict";

require( 'Unitest' ).enable();

if ( process.argv[2] == 'nocolor' ) {
	Unitest.noColor();
}

require( './FormatUtils.js' );
require( './TimeUtils.js' );

tests/TimeUtils.js
"use strict";

var TimeUtils = require( '../TimeUtils' );

Unitest( 'TimeUtils.timeAgo()', function ( test ) {

	test.eq( TimeUtils.timeAgoFormat( 60*1000 ), '1 minute' );
	test.eq( TimeUtils.timeAgoFormat( 60*1000+30*1000 ), '1 minute 30 seconds' );
	test.eq( TimeUtils.timeAgoFormat( 30*1000+500 ), '30 seconds 500 milliseconds' );
	test.eq( TimeUtils.timeAgoFormat( 30*1000+500, { LongFormat: '%d%s%d%s', Strings: 'short' } ), '30s500ms' );

} );

tests/FormatUtils.js
"use strict";

var FormatUtils = require( '../FormatUtils' );

Unitest( 'FormatUtils.formatBytes()', function ( test ) {

	test.eq( FormatUtils.formatBytes( 1000*1000 ), '1.00 MB' );
	test.eq( FormatUtils.formatBytes( 1024*1024, { UseSi: false, ForceUnit: 'KiB' } ), '1024.00 KiB' );

} );

Stats.js
"use strict";

//ported from https://github.com/Perennials/providerkit-core-php/blob/master/src/utils/Debug.php

var TimeUtils = require( './TimeUtils' );
var FormatUtils = require( './FormatUtils' );

class Stats {

	constructor () {
		this._timerStack = [];
		this._timers = new WeakMap();
		this._timers2 = {};
		this._stats = {};
		this._counts = {};
		this._units = {};
	}

	startTimer ( name ) {
		var time = Date.now();
		if ( name instanceof Object ) {
			this._timers.set( name, time );
		}
		else if ( name ) {
			this._timers2[ name ] = time;
		}
		else {
			this._timerStack.push( time );
		}
		return time;
	}

	endTimer ( name ) {
		var time = Date.now();
		if ( name instanceof Object ) {
			return time - this._timers.get( name );
		}
		else if ( name ) {
			return time - this._timers2[ name ];
		}
		else {
			return time - this._timerStack.pop();
		}
	}

	saveTimer ( name, stat ) {
		var t = undefined;
		if ( stat !== undefined ) {
			t = this.endTimer( name );
		}
		else {
			stat = name;
			t = this.endTimer();
		}
		this.setStat( stat, t, 'ms' );
		return t;
	}

	addTimer ( name, stat ) {
		var t = undefined;
		if ( stat !== undefined ) {
			t = this.endTimer( name );
		}
		else {
			stat = name;
			t = this.endTimer();
		}
		this.addStat( stat, t, 'ms' );
		return t;
	}

	setStat ( stat, value, unit ) {
		this._stats[ stat ] = value;
		this._counts[ stat ] = 1;
		if ( unit !== undefined ) {
			this._units[ stat ] = unit;
		}
		return value;
	}

	addStat ( stat, value, unit ) {
		if ( this._stats[ stat ] === undefined ) {
			this._stats[ stat ] = value;
			this._counts[ stat ] = 1;
		}
		else {
			this._stats[ stat ] += value;
			++this._counts[ stat ];
		}
		if ( unit !== undefined ) {
			this._units[ stat ] = unit;
		}
		return this._stats[ stat ];
	}

	getStat ( stat ) {
		return this._stats[ stat ];
	}

	setUnit ( stat, unit ) {
		return this._units[ stat ] = unit;
	}

	getUnit ( stat ) {
		return this._units[ stat ];
	}

	getCount ( stat ) {
		return this._counts[ stat ] || 0;
	}

	getAverage ( stat ) {
		v = this.getStat( stat );
		n = this.getCount( stat );
		if ( n > 0 ) {
			return v / n;
		}
		else {
			return undefined;
		}
	}

	saveAverage ( stat, newname ) {
		a = this.getAverage( stat );
		if ( a !== undefined ) {
			return this.setStat( newname, a, this.getUnit( stat ) );
		}
		else {
			return undefined;
		}
	}

	getStats ( options ) {
		options = options || Stats.DefaultFormatting;
		var timeOptions = options.Time || Stats.DefaultFormatting.Time;
		var byteOptions = options.Byte || Stats.DefaultFormatting.Byte;
		var ret = {};
		var stats = this._stats;
		for ( var k in stats ) {
			var v = stats[ k ];
			var u = this.getUnit( k );
			if ( u == 'b' ) {
				v = FormatUtils.formatBytes( v, byteOptions );
			}
			else if ( u == 'ms' ) {
				if ( v > 6000 ) {
					v = TimeUtils.timeAgoFormat( v, timeOptions );
				}
				else {
					v = ( v / 1000 ) + ' s';
				}
			}
			else if ( u !== undefined ) {
				v = v + ' ' + u;
			}
			ret[ k ] = v;
		}
		var keys = Object.keys( ret ).sort();
		var sortedRet = {};
		for ( var i = 0, iend = keys.length; i < iend; ++i ) {
			var key = keys[ i ];
			sortedRet[ key ] = ret[ key ];
		}

		return sortedRet;
	}

}

Stats.static( {
	DefaultFormatting: {
		Bytes: undefined,
		Time: { Precise: true }
	}
} );

module.exports = Stats;

FormatUtils.js
"use strict";

require( 'Prototype' );
var Sprintf = require( 'sprintf-js' ).sprintf;

//ported from https://github.com/Perennials/providerkit-core-php/blob/master/src/utils/FormatUtils.php

class FormatUtils {

	static formatBytes ( bytes, options ) {

		if ( !Object.isObject( options ) ) {
			options = {};
		}
		// Format string
		var format = options.Format;
		if ( !String.isString( options.Format ) ) {
			format = FormatUtils.DefaultFormat.Bytes;
		}

		var si = options.UseSi !== undefined ? options.UseSi : true;
		var forceunit = options.ForceUnit;

		// IEC prefixes (binary)
		if ( si == false || (String.isString( forceunit ) && forceunit.indexOf( 'i' ) >= 0) ) {
			var units = [ 'B', 'KiB', 'MiB', 'GiB', 'TiB', 'PiB' ];
			var mod = 1024;
		}
		// SI prefixes (decimal)
		else {
			var units = [ 'B', 'KB', 'MB', 'GB', 'TB', 'PB' ];
			var mod = 1000;
		}

		// Determine unit to use
		var power = 0;
		if ( (power = units.indexOf( forceunit )) < 0 ) {
			power = (bytes > 0) ? Math.floor( Math.log( bytes ) / Math.log( mod ) ) : 0;
		}

		return Sprintf( format, bytes / Math.pow( mod, power ), units[ power ] );
	}
}

FormatUtils.static( {
	DefaultFormat: {
		Bytes: '%01.2f %s'
	}
} )

module.exports = FormatUtils;

TimeUtils.js
"use strict";

require( 'Prototype' );
var Sprintf = require( 'sprintf-js' ).sprintf;

//ported from https://github.com/Perennials/providerkit-core-php/blob/master/src/utils/TimeUtils.php

const MillisecsPerSecond = 1000;
const MillisecsPerMinute = 60 * MillisecsPerSecond;
const MillisecsPerHour = 60 * MillisecsPerMinute;
const MillisecsPerDay = 24 * MillisecsPerHour;

var DefStrings = {
	Long: {
		year: 'year',
		years: 'years',
		month: 'month',
		months: 'months',
		day: 'day',
		days: 'days',
		hour: 'hour',
		hours: 'hours',
		minute: 'minute',
		minutes: 'minutes',
		second: 'second',
		seconds: 'seconds',
		millisecond: 'millisecond',
		milliseconds: 'milliseconds',
	},
	Short: {
		year: 'y',
		years: 'y',
		month: 'm',
		months: 'm',
		day: 'd',
		days: 'd',
		hour: 'h',
		hours: 'h',
		minute: 'min',
		minutes: 'min',
		second: 's',
		seconds: 's',
		millisecond: 'ms',
		milliseconds: 'ms',
	}
};

var DefFormat = {
	Short: '%d %s',
	Long: '%d %s %d %s'
};

class TimeUtils {

	static timeAgoFormat ( date2, date1, options ) {

		if ( date2 instanceof Date ) {
			data2 = date2.valueOf();
		}
		
		if ( date1 instanceof Date ) {
			date1 = date1.valueOf();
		}

		if ( date1 instanceof Object ) {
			options = date1;
			date1 = null;
		}

		var precise = options ? options.Precise || true : true;
		var format_long = options ? options.LongFormat || DefFormat.Long : DefFormat.Long;
		var format_short = options ? options.ShortFormat || DefFormat.Short : DefFormat.Short;
		var strings = options ? options.Strings || DefStrings.Long : DefStrings.Long;
		if ( strings == 'short' ) {
			strings = DefStrings.Short;
		}
		var r = TimeUtils.timeAgo( date2, date1, { Precise: precise } );
		if ( !precise || r.length < 4 ) {
			return Sprintf( format_short, r[ 0 ], strings[ r[ 1 ] ] );
		}
		else {
			return Sprintf( format_long, r[ 0 ], strings[ r[ 1 ] ], r[ 2 ], strings[ r[ 3 ] ] );
		}
	}

	static timeAgo ( date2, date1, options ) {

		if ( date2 instanceof Date ) {
			data2 = date2.valueOf();
		}

		if ( date1 instanceof Date ) {
			date1 = date1.valueOf();
		}

		var precise = (options instanceof Object && options.Precise) || true;
		var year2 = new Date( date2 ).getYear();
		var month2 = new Date( date2 ).getMonth();
		if ( !(date1 > 0) ) {
			var year1 = year2;
			var month1 = month2;
			var date1 = 0;
		}
		else {
			var year1 = new Date( date1 ).getYear();
			var month1 = new Date( date1 ).getMonth();
		}
		var years = year2 - year1;
		var months = years > 0 ? 12 - month1 + month2 : month2 - month1;
		if ( months >= 12 ) {
			++years;
			months -= 12;
		}

		var i1 = 0;
		var n1 = null;
		var i2 = 0;
		var n2 = null;

		if ( years >= 1 ) {
			i1 = years;
			i2 = months;
			n1 = 'year';
			n2 = 'month';
		}
		else {

			var diff = date2 - date1;

			var days = Math.floor( diff / MillisecsPerDay );

			if ( months >= 1 && days >= 28 ) {
				//months' length vary so we cant calculate the days without calendar knowledge
				i1 = months;
				n1 = 'month';
			}
			else {
				if ( days >= 1 ) {
					var rest = diff % MillisecsPerDay;
					var hours = Math.floor( rest / MillisecsPerHour );
					i1 = days;
					i2 = hours;
					n1 = 'day';
					n2 = 'hour';
				}
				else {
					var hours = Math.floor( diff / MillisecsPerHour );
					if ( hours >= 1 ) {
						var rest = diff % MillisecsPerHour;
						var minutes = Math.floor( rest / MillisecsPerMinute );
						i1 = hours;
						i2 = minutes;
						n1 = 'hour';
						n2 = 'minute';
					}
					else {
						var minutes = Math.floor( diff / MillisecsPerMinute );
						if ( minutes >= 1 ) {
							var rest = diff % MillisecsPerMinute;
							var seconds = Math.floor( rest / MillisecsPerSecond );
							i1 = minutes;
							i2 = seconds;
							n1 = 'minute';
							n2 = 'second';
						}
						else {
							var seconds = Math.floor( diff / MillisecsPerSecond );
							if ( seconds >= 1 ) {
								var rest = diff % MillisecsPerSecond;
								var milliseconds = rest;
								i1 = seconds;
								i2 = milliseconds;
								n1 = 'second';
								n2 = 'millisecond';
							}
							else {
								i1 = diff;
								if ( i1 < 1 ) {
									i1 = 1;
								}
								n1 = 'millisecond';
							}
						}
					}
				}
			}
		}

		if ( i1 > 1 ) {
			n1 += 's';
		}
		if ( i2 > 1 ) {
			n2 += 's';
		}

		if ( !precise || i2 == 0 ) {
			return [ i1, n1 ];
		}
		else {
			return [ i1, n1, i2, n2 ];
		}
	}
}

TimeUtils.static( {
	DefaultStrings: DefStrings,
	DefaultFormat: DefFormat
} );

module.exports = TimeUtils;